sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/TypeUtils.scala

```diff
@@ -73,4 +73,14 @@ object TypeUtils {
     }
     x.length - y.length
   }
+
+  /**
+   * Returns true if elements of the data type could be used as items of a hash set or as keys
+   * of a hash map.
+   */
+  def typeCanBeHashed(dataType: DataType): Boolean = dataType match {
```

Comment:
hey, this is a weird name, `byte[]` can also be hashed. I'd rather call it `typeWithProperEquals`, and document it as @mgaido91 proposed. I don't think we need to consider `hashCode` here, it's a rule in java world that equals and hashCode should be defined in a coherent way.

Reply:
I will change it :)

Just one question to ```hashCode```. If ```case classes``` are used, ```equals``` and ```hashCode``` are generated by compiler. But if we define ```equals``` manually, shouldn't also hold ```a.equals(b) == true``` => ```a.hashCode == b.hashCode```?

Reply:
we should, and I believe Spark has enforced it with style checker: when you override `equals`, you must override `hashCode` too.

Reply:
Ok, thanks!